### ISSUE TITLE
Failed to import EB_q easyblock

### ISSUE THREAD

**Issue by idanbudin** [easybuilders/easybuild-framework#2383]
Trying to install bcl2fast via eb with this: [](https://github.com/verdurin/easybuild-easyconfigs-devel/blob/master/bcl2fastq2-2.17.1.14-foss-2016a.eb).

I have used the following command line : `eb /home/me/Downloads/bcl2fastq2-2.17.1.14-foss-2016a.eb --robot.`
Which caused this error output:

```
== temporary log file in case of crash /tmp/eb-VEpgli/easybuild-zr8Caj.log
ERROR: Failed to process easyconfig /home/me/Downloads/bcl2fastq2-2.17.1.14-foss-2016a.eb: Failed to import EB_q easyblock: Failed to import EB_q from easybuild.easyblocks.q: 'module' object has no attribute 'EB_q'
```

**Any idea what is the problem and what can I do to fix it?**

The log file is below:

```
== 2018-01-22 17:50:54,365 main.py:79 INFO This is EasyBuild 3.5.1 (framework: 3.5.1, easyblocks: 3.5.1) on host IdanLinux.
== 2018-01-22 17:50:54,365 main.py:82 INFO Command line: --robot='/home/me/software/EasyBuild/3.5.1/lib/python2.7/site-packages/easybuild_easyconfigs-3.5.1-py2.7.egg/easybuild/easyconfigs' /home/me/Downloads/bcl2fastq2-2.17.1.14-foss-2016a.eb
== 2018-01-22 17:50:54,365 main.py:84 INFO Using /tmp/eb-VEpgli as temporary directory
== 2018-01-22 17:50:54,365 robot.py:56 INFO Using robot path(s): ['/home/me/software/EasyBuild/3.5.1/lib/python2.7/site-packages/easybuild_easyconfigs-3.5.1-py2.7.egg/easybuild/easyconfigs']
== 2018-01-22 17:50:54,366 options.py:1199 INFO Using default external modules metadata cfg files: ['/home/me/software/EasyBuild/3.5.1/lib/python2.7/site-packages/easybuild_framework-3.5.1-py2.7.egg/etc/cray_external_modules_metadata.cfg']
== 2018-01-22 17:50:54,367 environment.py:97 INFO Environment variable LMOD_QUIET set to 1 (previously undefined)
== 2018-01-22 17:50:54,368 environment.py:97 INFO Environment variable LMOD_IGNORE_CACHE set to 1 (previously undefined)
== 2018-01-22 17:50:54,368 environment.py:97 INFO Environment variable LMOD_REDIRECT set to no (previously undefined)
== 2018-01-22 17:50:54,368 filetools.py:343 INFO Command lmod found at /home/me/lmod/6.1/libexec/lmod
== 2018-01-22 17:50:54,368 filetools.py:343 INFO Command lmod found at /home/me/lmod/6.1/libexec/lmod
== 2018-01-22 17:50:54,368 filetools.py:343 INFO Command lmod found at /home/me/lmod/6.1/libexec/lmod
== 2018-01-22 17:50:54,368 modules.py:268 INFO Full path for module command is /home/me/lmod/6.1/libexec/lmod, so using it
== 2018-01-22 17:50:54,390 environment.py:97 INFO Environment variable LD_LIBRARY_PATH set to  (previously undefined)
== 2018-01-22 17:50:54,391 environment.py:97 INFO Environment variable LD_PRELOAD set to  (previously undefined)
== 2018-01-22 17:50:54,391 modules.py:399 INFO Prepended list of module paths with path used by EasyBuild: /home/me/.local/easybuild/modules/all
== 2018-01-22 17:50:54,391 run.py:183 INFO running cmd: type module 
== 2018-01-22 17:50:54,412 modules.py:227 INFO Found version 6.1
== 2018-01-22 17:50:54,412 modules.py:235 INFO Converted actual version to '6.1'
== 2018-01-22 17:50:54,436 environment.py:97 INFO Environment variable LD_LIBRARY_PATH set to  (previous value: '')
== 2018-01-22 17:50:54,436 environment.py:97 INFO Environment variable LD_PRELOAD set to  (previous value: '')
== 2018-01-22 17:50:54,436 filetools.py:343 INFO Command eb found at /home/me/software/EasyBuild/3.5.1/bin/eb
== 2018-01-22 17:50:54,438 environment.py:97 INFO Environment variable LMOD_QUIET set to 1 (previous value: '1')
== 2018-01-22 17:50:54,438 environment.py:97 INFO Environment variable LMOD_IGNORE_CACHE set to 1 (previous value: '1')
== 2018-01-22 17:50:54,438 environment.py:97 INFO Environment variable LMOD_REDIRECT set to no (previous value: 'no')
== 2018-01-22 17:50:54,438 filetools.py:343 INFO Command lmod found at /home/me/lmod/6.1/libexec/lmod
== 2018-01-22 17:50:54,438 filetools.py:343 INFO Command lmod found at /home/me/lmod/6.1/libexec/lmod
== 2018-01-22 17:50:54,438 filetools.py:343 INFO Command lmod found at /home/me/lmod/6.1/libexec/lmod
== 2018-01-22 17:50:54,438 modules.py:268 INFO Full path for module command is /home/me/lmod/6.1/libexec/lmod, so using it
== 2018-01-22 17:50:54,438 modules.py:399 INFO Prepended list of module paths with path used by EasyBuild: /home/me/.local/easybuild/modules/all
== 2018-01-22 17:50:54,438 easyconfig.py:283 INFO Obtained list of valid module classes: ['base', 'bio', 'cae', 'chem', 'compiler', 'data', 'debugger', 'devel', 'geo', 'ide', 'lang', 'lib', 'math', 'mpi', 'numlib', 'perf', 'phys', 'system', 'toolchain', 'tools', 'vis']
== 2018-01-22 17:50:54,452 build_log.py:158 ERROR EasyBuild crashed with an error (at ?:124 in __init__): Failed to import EB_q easyblock: Failed to import EB_q from easybuild.easyblocks.q: 'module' object has no attribute 'EB_q' (at easybuild/framework/easyconfig/easyconfig.py:1135 in get_easyblock_class)
== 2018-01-22 17:50:54,452 build_log.py:158 ERROR EasyBuild crashed with an error (at ?:124 in __init__): Failed to process easyconfig /home/me/Downloads/bcl2fastq2-2.17.1.14-foss-2016a.eb: Failed to import EB_q easyblock: Failed to import EB_q from easybuild.easyblocks.q: 'module' object has no attribute 'EB_q' (at easybuild/framework/easyconfig/easyconfig.py:1263 in process_easyconfig)
```

**Comment by boegel** [comment]
@idanbudin Hmm, this is a weird issue... I'm lacking the required info to pinpoint the actual issue.

* Which EasyBuild version are you using?
* What's your EasyBuild configuration (output of `eb --show-config`)?
* Can you provide a debug log, collected via `eb --debug bcl2fastq2-2.17.1.14-foss-2016a.eb --robot`?
* Can you share the `bcl2fastq2-2.17.1.14-foss-2016a.eb` easyconfig file you are using?
* Do you have a `q.py` module available in the local directory that EasyBuild could be picking up?

**Comment by idanbudin** [comment]
- The version is: 3.5.1

- My configuration:
buildpath      (D) = /home/me/.local/easybuild/build
installpath    (D) = /home/me/.local/easybuild
repositorypath (D) = /home/me/.local/easybuild/ebfiles_repo
robot-paths    (D) = /home/me/software/EasyBuild/3.5.1/lib/python2.7/site-packages/easybuild_easyconfigs-3.5.1-py2.7.egg/easybuild/easyconfigs
sourcepath     (D) = /home/me/.local/easybuild/sources

-The debug log is attached by:
[easybuild-tkF_0Q.log](https://github.com/easybuilders/easybuild-framework/files/1659454/easybuild-tkF_0Q.log)
.

- I though that I have shared the eb file, but apparently I did not so here it is: [https://github.com/verdurin/easybuild-easyconfigs-devel/blob/master/bcl2fastq2-2.17.1.14-foss-2016a.eb](url)

- I couldn't find any file with that name on my system

Thanks!

**Comment by idanbudin** [comment]
I've managed to install bcl2fastq with all its dependencies without easybuild thanks to the following link:
https://github.com/igorbarinov/bcl2fastq/blob/master/install-2.17.sh
So for now this problem doesn't bother me anymore.
Thank you anyway!

**Comment by tardigradus** [comment]
Using a fairly fresh install of Centos 7 with EasyBuild 3.6.1 and Lmod 6.5.1, I have experienced the same error with the following recipe: 

https://github.com/easybuilders/easybuild-easyconfigs/blob/master/easybuild/easyconfigs/r/R/R-3.4.4-foss-2018a-X11-20180131.eb

My configuration is:

buildpath      (D) = /home/xxx/.local/easybuild/build
containerpath  (D) = /home/xxx/.local/easybuild/containers
installpath    (D) = /home/xxx/.local/easybuild
repositorypath (D) = /home/xxx/.local/easybuild/ebfiles_repo
robot-paths    (D) = /home/xxx/EasyBuild/software/EasyBuild/3.6.1/lib/python2.7/site-packages/easybuild_easyconfigs-3.6.1-py2.7.egg/easybuild/easyconfigs
sourcepath     (D) = /home/xxx/.local/easybuild/sources

Here is the log file:

[easybuild-oYDBD0.log](https://github.com/easybuilders/easybuild-framework/files/2140382/easybuild-oYDBD0.log)

**Comment by tardigradus** [comment]
OK, the problem was that I goofed while downloading the eb-file from Github.  Instead of getting the raw version of the file, I downloaded the HTML version of the page like this:

`wget https://github.com/easybuilders/easybuild-easyconfigs/blob/master/easybuild/easyconfigs/r/R/R-3.4.4-foss-2018a-X11-20180131.eb`

Using this as input made EasyBuild barf.  However, once I got the raw version via

`wget https://raw.githubusercontent.com/easybuilders/easybuild-easyconfigs/master/easybuild/easyconfigs/r/R/R-3.4.4-foss-2018a-X11-20180131.eb`

all was good.

**Comment by boegel** [comment]
@tardigradus Interesting!

I'm indeed able to get the `Failed to import EB_q easyblock` by downloading the easyconfig file via `wget`.

We clearly need to do a better job of validating that whatever is being passed is actually an easyconfig file...

**Comment by boegel** [comment]
Here's the culprit:

```
$ grep "^[ ]*name" R-3.4.4-foss-2018a-X11-20180131.eb
          name="q"
```

That gets picked up as a software name specification, and with no `easyblock` spec being found in this "easyconfig file", EasyBuild continues to try and find a `EB_q` easyblock to use to install the software named `q` with...

I'll take a look at fixing this and reporting that whatever is provided doesn't seem to be an easyconfig file.

**Comment by tardigradus** [comment]
Great, thanks!  I really should have known better, because I have make the same mistake in the past.  However, it would be unfortunate if people like the OP gave up on EasyBuild, just becase the input file was broken.

**Comment by boegel** [comment]
Fully agreed, I'm working on a fix to recognize that the provided input file is not an easyconfig file (but that's less trivial than it seems, because of a chicking-egg situation w.r.t. figuring out the easyblock that should be used...)

**Comment by boegel** [comment]
Turns out it's actually easier than I thought it would be, see fix in #2523

**Comment by tardigradus** [comment]
Thanks, that's great.  With the error message produced by the fix I would have certainly checked the contents of the file sooner.

**Comment by davidliu-CAS** [comment]
Hi boegel. I ran into the same error today on RHEL 7.2 with EasyBuild 3.6.1 and Lmod 6.5.1.
The system was freshly installed a couple of days ago. Pacakges ESMF and Trilinos were successfully built by easybuild under root. The last successful build was Trilinos/12.4.2-intel-2015b-python-2.7.11 about 20 hours ago.  I have experienced the same error today at dryrun step for every package I have tried even for update using--install-latest-eb-release. The same error occurred for user too. I didn't download any easyconfig file but only deleted the /root/.local folder yesterday to save some space. Would it cause the problem? Thanks!

My configuration is:

allow-use-as-root-and-accept-consequences (C) = True
buildpath                                 (D) = /root/.local/easybuild/build
containerpath                             (D) = /root/.local/easybuild/containers
installpath                               (D) = /root/.local/easybuild
repositorypath                            (D) = /root/.local/easybuild/ebfiles_repo
robot-paths                               (D) = /public/software/easybuild/software/EasyBuild/3.6.1/lib/python2.7/site-packages/easybuild_easyconfigs-3.6.1-py2.7.egg/easybuild/easyconfigs
sourcepath                                (D) = /root/.local/easybuild/sources


and  the log file for update using--install-latest-eb-release --allow-use-as-root-and-accept-consequences

== 2018-06-29 13:55:56,106 main.py:80 INFO This is EasyBuild 3.6.1 (framework: 3.6.1, easyblocks: 3.6.1) on host admin.
== 2018-06-29 13:55:56,106 main.py:83 INFO Command line: --allow-use-as-root-and-accept-consequences --install-latest-eb-release
== 2018-06-29 13:55:56,106 main.py:85 INFO Using /tmp/eb-tZdCZb as temporary directory
== 2018-06-29 13:55:56,106 robot.py:56 INFO Using robot path(s): ['/public/software/easybuild/software/EasyBuild/3.6.1/lib/python2.7/site-packages/easybuild_easyconfigs-3.6.1-py2.7.egg/easybuild/easyconfigs']
== 2018-06-29 13:55:56,106 options.py:1239 INFO Using default external modules metadata cfg files: ['/public/software/easybuild/software/EasyBuild/3.6.1/lib/python2.7/site-packages/easybuild_framework-3.6.1-py2.7.egg/etc/cray_external_modules_metadata.cfg', '/public/software/easybuild/software/EasyBuild/3.6.1/lib/python2.7/site-packages/easybuild_framework-3.6.1-py2.7.egg/etc/ohpc_external_modules_metadata.cfg']
== 2018-06-29 13:55:56,113 environment.py:97 INFO Environment variable LMOD_QUIET set to 1 (previously undefined)
== 2018-06-29 13:55:56,113 environment.py:97 INFO Environment variable LMOD_IGNORE_CACHE set to 1 (previously undefined)
== 2018-06-29 13:55:56,113 environment.py:97 INFO Environment variable LMOD_REDIRECT set to no (previously undefined)
== 2018-06-29 13:55:56,113 filetools.py:351 WARNING Could not find command 'lmod' (with permissions to read/execute it) in $PATH (['/public/software/easybuild/software/EasyBuild/3.6.1/bin', '/usr/local/bin', '/usr/local/sbin', '/opt/clusconf/bin', '/usr/local/sbin', '/usr/local/bin', '/usr/sbin', '/usr/bin', '/root/bin'])
== 2018-06-29 13:55:56,113 filetools.py:343 INFO Command /usr/share/lmod/lmod/libexec/lmod found at /usr/share/lmod/lmod/libexec/lmod
== 2018-06-29 13:55:56,113 modules.py:264 INFO Full path for module command is /usr/share/lmod/lmod/libexec/lmod, so using it
== 2018-06-29 13:55:56,184 environment.py:97 INFO Environment variable LD_LIBRARY_PATH set to  (previously undefined)
== 2018-06-29 13:55:56,184 environment.py:97 INFO Environment variable LD_PRELOAD set to  (previously undefined)
== 2018-06-29 13:55:56,185 modules.py:409 INFO Prepended list of module paths with path used by EasyBuild: /root/.local/easybuild/modules/all
== 2018-06-29 13:55:56,255 environment.py:97 INFO Environment variable LD_LIBRARY_PATH set to  (previous value: '')
== 2018-06-29 13:55:56,255 environment.py:97 INFO Environment variable LD_PRELOAD set to  (previous value: '')
== 2018-06-29 13:55:56,325 environment.py:97 INFO Environment variable LD_LIBRARY_PATH set to  (previous value: '')
== 2018-06-29 13:55:56,325 environment.py:97 INFO Environment variable LD_PRELOAD set to  (previous value: '')
== 2018-06-29 13:55:56,397 environment.py:97 INFO Environment variable LD_LIBRARY_PATH set to  (previous value: '')
== 2018-06-29 13:55:56,397 environment.py:97 INFO Environment variable LD_PRELOAD set to  (previous value: '')
== 2018-06-29 13:55:56,468 environment.py:97 INFO Environment variable LD_LIBRARY_PATH set to  (previous value: '')
== 2018-06-29 13:55:56,469 environment.py:97 INFO Environment variable LD_PRELOAD set to  (previous value: '')
== 2018-06-29 13:55:56,539 environment.py:97 INFO Environment variable LD_LIBRARY_PATH set to  (previous value: '')
== 2018-06-29 13:55:56,539 environment.py:97 INFO Environment variable LD_PRELOAD set to  (previous value: '')
== 2018-06-29 13:55:56,539 modules.py:426 INFO $MODULEPATH set via list of module paths (w/ 'module use'): /root/.local/easybuild/modules/all:/public/software/easybuild/modules/all:/public/software/modules/all:/etc/modulefiles:/usr/share/modulefiles:/usr/share/modulefiles/Linux:/usr/share/modulefiles/Core:/usr/share/lmod/lmod/modulefiles/Core
== 2018-06-29 13:55:56,539 run.py:192 INFO running cmd: type module 
== 2018-06-29 13:55:56,571 modules.py:223 INFO Found version 6.5.1
== 2018-06-29 13:55:56,572 modules.py:231 INFO Converted actual version to '6.5.1'
== 2018-06-29 13:55:56,616 environment.py:97 INFO Environment variable LD_LIBRARY_PATH set to  (previous value: '')
== 2018-06-29 13:55:56,616 environment.py:97 INFO Environment variable LD_PRELOAD set to  (previous value: '')
== 2018-06-29 13:55:56,617 filetools.py:343 INFO Command eb found at /public/software/easybuild/software/EasyBuild/3.6.1/bin/eb
== 2018-06-29 13:55:56,617 filetools.py:1108 INFO Creating directory /tmp/eb-tZdCZb/tmpEKXETH/easybuilders (parents: True, set_gid: False, sticky: False)
== 2018-06-29 13:55:56,617 github.py:1418 WARNING No GitHub user name provided, required for fetching GitHub token.
== 2018-06-29 13:56:01,540 filetools.py:471 INFO Downloaded file develop.tar.gz from url https://github.com/easybuilders/easybuild-easyconfigs/archive/develop.tar.gz to /tmp/eb-tZdCZb/tmpEKXETH/easybuilders/develop.tar.gz
== 2018-06-29 13:56:01,540 filetools.py:491 INFO Successful download of file develop.tar.gz from url https://github.com/easybuilders/easybuild-easyconfigs/archive/develop.tar.gz to path /tmp/eb-tZdCZb/tmpEKXETH/easybuilders/develop.tar.gz
== 2018-06-29 13:56:01,541 run.py:192 INFO running cmd: tar xzf /tmp/eb-tZdCZb/tmpEKXETH/easybuilders/develop.tar.gz 
== 2018-06-29 13:56:01,920 environment.py:97 INFO Environment variable LMOD_QUIET set to 1 (previous value: '1')
== 2018-06-29 13:56:01,920 environment.py:97 INFO Environment variable LMOD_IGNORE_CACHE set to 1 (previous value: '1')
== 2018-06-29 13:56:01,920 environment.py:97 INFO Environment variable LMOD_REDIRECT set to no (previous value: 'no')
== 2018-06-29 13:56:01,920 filetools.py:351 WARNING Could not find command 'lmod' (with permissions to read/execute it) in $PATH (['/public/software/easybuild/software/EasyBuild/3.6.1/bin', '/usr/local/bin', '/usr/local/sbin', '/opt/clusconf/bin', '/usr/local/sbin', '/usr/local/bin', '/usr/sbin', '/usr/bin', '/root/bin'])
== 2018-06-29 13:56:01,921 filetools.py:343 INFO Command /usr/share/lmod/lmod/libexec/lmod found at /usr/share/lmod/lmod/libexec/lmod
== 2018-06-29 13:56:01,921 modules.py:264 INFO Full path for module command is /usr/share/lmod/lmod/libexec/lmod, so using it
== 2018-06-29 13:56:01,921 modules.py:409 INFO Prepended list of module paths with path used by EasyBuild: /root/.local/easybuild/modules/all
== 2018-06-29 13:56:01,992 environment.py:97 INFO Environment variable LD_LIBRARY_PATH set to  (previous value: '')
== 2018-06-29 13:56:01,992 environment.py:97 INFO Environment variable LD_PRELOAD set to  (previous value: '')
== 2018-06-29 13:56:02,063 environment.py:97 INFO Environment variable LD_LIBRARY_PATH set to  (previous value: '')
== 2018-06-29 13:56:02,063 environment.py:97 INFO Environment variable LD_PRELOAD set to  (previous value: '')
== 2018-06-29 13:56:02,134 environment.py:97 INFO Environment variable LD_LIBRARY_PATH set to  (previous value: '')
== 2018-06-29 13:56:02,135 environment.py:97 INFO Environment variable LD_PRELOAD set to  (previous value: '')
== 2018-06-29 13:56:02,206 environment.py:97 INFO Environment variable LD_LIBRARY_PATH set to  (previous value: '')
== 2018-06-29 13:56:02,207 environment.py:97 INFO Environment variable LD_PRELOAD set to  (previous value: '')
== 2018-06-29 13:56:02,277 environment.py:97 INFO Environment variable LD_LIBRARY_PATH set to  (previous value: '')
== 2018-06-29 13:56:02,277 environment.py:97 INFO Environment variable LD_PRELOAD set to  (previous value: '')
== 2018-06-29 13:56:02,277 modules.py:426 INFO $MODULEPATH set via list of module paths (w/ 'module use'): /root/.local/easybuild/modules/all:/public/software/easybuild/modules/all:/public/software/modules/all:/etc/modulefiles:/usr/share/modulefiles:/usr/share/modulefiles/Linux:/usr/share/modulefiles/Core:/usr/share/lmod/lmod/modulefiles/Core
== 2018-06-29 13:56:02,277 easyconfig.py:282 INFO Obtained list of valid module classes: ['base', 'bio', 'cae', 'chem', 'compiler', 'data', 'debugger', 'devel', 'geo', 'ide', 'lang', 'lib', 'math', 'mpi', 'numlib', 'perf', 'phys', 'system', 'toolchain', 'tools', 'vis']
== 2018-06-29 13:56:02,299 build_log.py:158 ERROR EasyBuild crashed with an error (at easybuild/software/EasyBuild/3.6.1/lib/python2.7/site-packages/vsc_base-2.7.2-py2.7.egg/vsc/utils/exceptions.py:124 in __init__): Failed to obtain class for EB_EasyBuildMeta easyblock (not available?): No module named easybuildmeta (at easybuild/software/EasyBuild/3.6.1/lib/python2.7/site-packages/easybuild_framework-3.6.1-py2.7.egg/easybuild/framework/easyconfig/easyconfig.py:1143 in get_easyblock_class)
== 2018-06-29 13:56:02,299 build_log.py:158 ERROR EasyBuild crashed with an error (at easybuild/software/EasyBuild/3.6.1/lib/python2.7/site-packages/vsc_base-2.7.2-py2.7.egg/vsc/utils/exceptions.py:124 in __init__): Failed to process easyconfig /tmp/eb-tZdCZb/tmpEKXETH/easybuilders/easybuild-easyconfigs-develop/easybuild/easyconfigs/e/EasyBuild/EasyBuild-3.6.1.eb: Failed to obtain class for EB_EasyBuildMeta easyblock (not available?): No module named easybuildmeta (at easybuild/software/EasyBuild/3.6.1/lib/python2.7/site-packages/easybuild_framework-3.6.1-py2.7.egg/easybuild/framework/easyconfig/easyconfig.py:1260 in process_easyconfig)

**Comment by boegel** [comment]
@davidliu-CAS The problem may look similar, but it's pretty different.

Can you still find `easybuildmeta.py` in your EasyBuild installation (in `/public/software/easybuild/software/EasyBuild/3.6.1`)?

It seems like that file is still there but empty...

**Comment by davidliu-CAS** [comment]
thanks, boegel. The file locates at /public/software/easybuild/lib/python2.7/site-packages/easybuild/easyblocks/e/easybuildmeta.py and is not empty.

# #
# Copyright 2013 Ghent University
#
# This file is part of EasyBuild,
# originally created by the HPC team of Ghent University (http://ugent.be/hpc/en),
# with support of Ghent University (http://ugent.be/hpc),
# the Flemish Supercomputer Centre (VSC) (https://www.vscentrum.be),
# Flemish Research Foundation (FWO) (http://www.fwo.be/en)
# and the Department of Economy, Science and Innovation (EWI) (http://www.ewi-vlaanderen.be/en).
#
# https://github.com/easybuilders/easybuild
#
# EasyBuild is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation v2.
#
# EasyBuild is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with EasyBuild.  If not, see <http://www.gnu.org/licenses/>.
# #
"""
EasyBuild support for installing EasyBuild, implemented as an easyblock

@author: Kenneth Hoste (UGent)
"""
import copy
import os
import re
from distutils.version import LooseVersion

from easybuild.easyblocks.generic.pythonpackage import PythonPackage
from easybuild.tools.build_log import EasyBuildError
from easybuild.tools.filetools import read_file
from easybuild.tools.modules import get_software_root_env_var_name
from easybuild.tools.ordereddict import OrderedDict
from easybuild.tools.utilities import flatten


# note: we can't use EB_EasyBuild as easyblock name, as that would require an easyblock named 'easybuild.py',
#       which would screw up namespacing and create all kinds of problems (e.g. easyblocks not being found anymore)
class EB_EasyBuildMeta(PythonPackage):
    """Support for install EasyBuild."""

    def __init__(self, *args, **kwargs):
        """Initialize custom class variables."""
        super(EB_EasyBuildMeta, self).__init__(*args, **kwargs)
        self.real_initial_environ = None

        self.easybuild_pkgs = ['easybuild-framework', 'easybuild-easyblocks', 'easybuild-easyconfigs']
        if LooseVersion(self.version) >= LooseVersion('2.0'):
            # deliberately include vsc-install & vsc-base twice;
            # first time to ensure the specified vsc-install/vsc-base package is available when framework gets installed
            self.easybuild_pkgs.insert(0, 'vsc-base')
            self.easybuild_pkgs.insert(0, 'vsc-install')
            # second time as last package to be installed, to ensure that the vsc-base version listed
            # in easy-install.pth is the one specified;
            # when installing the easybuild-* packages, the vsc-base version in easy-install.pth may be 'bumped'
            # if a newer vsc-base version is found somewhere (e.g. provided by the OS)
            self.easybuild_pkgs.extend(['vsc-base', 'vsc-install'])
            # consider setuptools first, in case it is listed as a sources
            self.easybuild_pkgs.insert(0, 'setuptools')

    def check_readiness_step(self):
        """Make sure EasyBuild can be installed with a loaded EasyBuild module."""
        env_var_name = get_software_root_env_var_name(self.name)
        if env_var_name in os.environ:
            os.environ.pop(env_var_name)
            self.log.debug("$%s is unset so EasyBuild can be installed with a loaded EasyBuild module" % env_var_name)
        else:
            self.log.debug("Not unsetting $%s since it's not set" % env_var_name)

        super(EB_EasyBuildMeta, self).check_readiness_step()

    def build_step(self):
        """No building for EasyBuild packages."""
        pass

    def install_step(self):
        """Install EasyBuild packages one by one."""
        try:
            subdirs = os.listdir(self.builddir)
            for pkg in self.easybuild_pkgs:
                seldirs = [x for x in subdirs if x.startswith(pkg)]
                if len(seldirs) != 1:
                    # setuptools is optional since it may be available in the OS;
                    # vsc-install and vsc-base sources are optional,
                    # they can be pulled in from PyPi when installing easybuild-framework too
                    if pkg not in ['setuptools', 'vsc-base', 'vsc-install']:
                        raise EasyBuildError("Failed to find required EasyBuild package %s (subdirs: %s, seldirs: %s)",
                                             pkg, subdirs, seldirs)

                else:
                    self.log.info("Installing package %s", pkg)
                    os.chdir(os.path.join(self.builddir, seldirs[0]))
                    super(EB_EasyBuildMeta, self).install_step()

        except OSError, err:
            raise EasyBuildError("Failed to install EasyBuild packages: %s", err)

    def post_install_step(self):
        """Remove setuptools.pth file that hard includes a system-wide (site-packages) path, if it is there."""

        setuptools_pth = os.path.join(self.installdir, self.pylibdir, 'setuptools.pth')
        if os.path.exists(setuptools_pth):
            setuptools_pth_txt = read_file(setuptools_pth)
            # any line that starts with '/' is a sign of trouble
            sys_path_regex = re.compile('^/', re.M)
            if sys_path_regex.search(setuptools_pth_txt):
                self.log.warning("Found %s, and includes one or more absolute system paths. Removing it.",
                                 setuptools_pth)
                try:
                    os.remove(setuptools_pth)
                except OSError as err:
                    raise EasyBuildError("Failed to remove %s: %s", setuptools_pth, err)

    def sanity_check_step(self):
        """Custom sanity check for EasyBuild."""

        # check whether easy-install.pth contains correct entries
        easy_install_pth = os.path.join(self.installdir, self.pylibdir, 'easy-install.pth')
        if os.path.exists(easy_install_pth):
            easy_install_pth_txt = read_file(easy_install_pth)
            for pkg in [p for p in self.easybuild_pkgs if p not in ['setuptools', 'vsc-install']]:
                if pkg == 'vsc-base':
                    # don't include strict version check for vsc-base
                    pkg_regex = re.compile(r"^\./%s" % pkg.replace('-', '_'), re.M)
                else:
                    major_minor_version = '.'.join(self.version.split('.')[:2])
                    pkg_regex = re.compile(r"^\./%s-%s" % (pkg.replace('-', '_'), major_minor_version), re.M)

                if not pkg_regex.search(easy_install_pth_txt):
                    raise EasyBuildError("Failed to find pattern '%s' in %s: %s",
                                         pkg_regex.pattern, easy_install_pth, easy_install_pth_txt)

        # list of dirs to check, by package
        # boolean indicates whether dir is expected to reside in Python lib/pythonX/site-packages dir
        subdirs_by_pkg = {
            'easybuild-framework': [('easybuild/framework', True), ('easybuild/tools', True)],
            'easybuild-easyblocks': [('easybuild/easyblocks', True)],
            'easybuild-easyconfigs': [('easybuild/easyconfigs', False)],
        }
        if LooseVersion(self.version) >= LooseVersion('2.0'):
            subdirs_by_pkg.update({
                'vsc-base': [('vsc/utils', True)],
            })

        # final list of directories to check, by setup tool
        # order matters, e.g. setuptools before distutils
        eb_dirs = OrderedDict()
        eb_dirs['setuptools'] = []
        eb_dirs['distutils.core'] = flatten([x for x in subdirs_by_pkg.values()])

        # determine setup tool (setuptools or distutils)
        setup_tool = None
        for tool in eb_dirs.keys():
            self.log.debug("Trying %s.." % tool)
            try:
                exec "from %s import setup" % tool
                del setup
                setup_tool = tool
                break
            except ImportError:
                pass
        self.log.debug('setup_tool: %s' % setup_tool)

        # for a setuptools installation, we need to figure out the egg dirs,
        # since we don't know the individual package versions
        if setup_tool == 'setuptools':
            try:
                installed_dirs = os.listdir(os.path.join(self.installdir, self.pylibdir))
                for (pkg, subdirs) in subdirs_by_pkg.items():
                    sel_dirs = [x for x in installed_dirs if x.startswith(pkg.replace('-', '_'))]
                    if not len(sel_dirs) == 1:
                        raise EasyBuildError("Failed to isolate installed egg dir for %s", pkg)

                    for (subdir, _) in subdirs:
                        # eggs always go in Python lib/pythonX/site-packages dir with setuptools
                        eb_dirs['setuptools'].append((os.path.join(sel_dirs[0], subdir), True))
            except OSError, err:
                raise EasyBuildError("Failed to determine sanity check dir paths: %s", err)

        # set of sanity check paths to check for EasyBuild
        custom_paths = {
            'files': ['bin/eb'],
            'dirs': [self.pylibdir] + [[x, os.path.join(self.pylibdir, x)][y] for (x, y) in eb_dirs[setup_tool]],
        }

        # make sure we don't trip over deprecated behavior in old EasyBuild versions
        eb_cmd = 'eb'
        if LooseVersion(self.version) <= LooseVersion('1.16.0'):
            eb_cmd = 'EASYBUILD_DEPRECATED=1.0 eb'

        # set of sanity check commands to run for EasyBuild
        custom_commands = [
            # this may spit out a wrong version, but that should be safe to ignore
            # occurs when the EasyBuild being used is newer than the EasyBuild being installed
            (eb_cmd, '--version'),
            (eb_cmd, '-a'),
            (eb_cmd, '-e ConfigureMake -a'),
        ]

        # (temporary) cleanse copy of initial environment to avoid conflict with (potentially) loaded EasyBuild module
        self.real_initial_environ = copy.deepcopy(self.initial_environ)
        for env_var in ['_LMFILES_', 'LOADEDMODULES']:
            if env_var in self.initial_environ:
                self.initial_environ.pop(env_var)
                os.environ.pop(env_var)
                self.log.debug("Unset $%s in current env and copy of original env to make sanity check work" % env_var)

        super(EB_EasyBuildMeta, self).sanity_check_step(custom_paths=custom_paths, custom_commands=custom_commands)

    def make_module_step(self, fake=False):
        """Create module file, before copy of original environment that was tampered with is restored."""
        modpath = super(EB_EasyBuildMeta, self).make_module_step(fake=fake)

        if not fake:
            # restore copy of original environment
            self.initial_environ = copy.deepcopy(self.real_initial_environ)
            self.log.debug("Restored copy of original environment")

        return modpath

**Comment by boegel** [comment]
@davidliu-CAS Hmm, that's exactly what it should be, so it's another problem.

Can you try and collect a debug log, using `eb --debug ...`?

**Comment by davidliu-CAS** [comment]
@boegel Thanks. Here you go the debug log~
[easybuild-NHObwD.log](https://github.com/easybuilders/easybuild-framework/files/2148460/easybuild-NHObwD.log)

**Comment by boegel** [comment]
@davidliu-CAS I suspect the problem you are seeing is a side-effect of running `eb` as root...

The debug info isn't detailed enough to really pinpoint the problem, but can you specify how you are running `eb` exactly? Are you loading the `EasyBuild` module as root, and then run the `eb` command? Or are you running it via `sudo eb`?

What does the following command produce when you run it as root?

```
python -c "import easybuild.easyblocks.easybuildmeta; print easybuild.easyblocks.easybuildmeta.__file__"
```

**Comment by boegel** [comment]
I'm closing this issue as the problem that was originally reported by @idanbudin has been fixed in #2523 (will be included in upcoming release EasyBuild v3.6.2).

@davidliu-CAS The problem you reported is another issue, so please open a separate issue for that with more details, so we can follow up on that.
